File: 2022/7/day7.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

int main()
{
    FILE *file = fopen("input/day7_input.txt", "r");
    int stack[20] = {0}, sizes[200] = {0};
    char line[30];
    while (fgets(line, 30, file))
    {
        if (!strcmp(line, "$ cd ..\n"))
        {
            int size = stack[stack[0]];
            stack[0]--;
            sizes[0]++;
            sizes[sizes[0]] = size;
            stack[stack[0]] += size;
        }
        else if (!strncmp(line, "$ cd ", 5))
        {
            stack[0]++;
            stack[stack[0]] = 0;
        }
        else if (isdigit(line[0]))
        {
            int size;
            sscanf(line, "%d", &size);
            stack[stack[0]] += size;
        }
    }
    
    while (stack[0] > 0)
    {
        sizes[0]++;
        sizes[sizes[0]] = stack[stack[0]];
        stack[0]--;
        if (stack[0] > 0)
            stack[stack[0]] += sizes[sizes[0]];
    }
    
    int sum = 0;
    int min = sizes[sizes[0]];
    for (int i = 1; i < sizes[0] + 1; i++)
    {
        if (sizes[i] <= 100000)
            sum += sizes[i];
        if (sizes[i] >= sizes[sizes[0]] - 40000000 && sizes[i] < min)
            min = sizes[i];
    }
    printf("Sol1: %d\n", sum);
    printf("Sol2: %d\n", min);

    return 0;
}

File: 2022/7/day7.py
with open("input/day7_input.txt") as f:
    input = f.read().splitlines()

stack, sizes = list(), list()

for line in input:
    if line == "$ cd ..":
        size = stack.pop()
        sizes.append(size)
        stack[-1] += size
    elif line[:5] == "$ cd ":
        stack.append(0)
    elif line[0].isdigit():
        stack[-1] += int(line.split()[0])

print(stack)

while stack:
    size = stack.pop()
    sizes.append(size)
    if stack:
        stack[-1] += size

print(sum(s for s in sizes if s <= 100000))
print(min(s for s in sizes if s >= sizes[-1] - 40000000))